### PR TITLE
Re-introduce concept of projection cache 'completion'

### PR DESCRIPTION
**Pull request by Aaron1011**
Instead of clearing out the cache entirely, we store
the intermediate evaluation result into the cache entry.
This accomplishes several things:

* We avoid the performance hit associated with re-evaluating
  the sub-obligations
* We avoid causing issues with incremental compilation, since
  the final evaluation result is always the same
* We avoid affecting other uses of the same `InferCtxt` which
  might care about 'side effects' from processing the sub-obligations
  (e,g. region constraints). Only code that is specifically aware
   of the new 'complete' code is affected

Authored to address issue  #89195

### CODE REVIEW

**Comment by nayato** [comment]
This fixes enormous regression we've seen introduced seemingly with https://github.com/rust-lang/rust/pull/85868.
Debug build before before #85868: 3 min
Same build after #85868: 12+ h - at which point we don't wait for it anymore and abort it.
Same build with this PR: 3.5 min

**Comment by Aaron1011** [comment]
@nayato: I'm glad to hear it! I came up with this as a fix for https://github.com/rust-lang/rust/issues/89195, which appears to have the same underlying cause as your issue.

Unfortunately, part of the issue with https://github.com/rust-lang/rust/issues/89195 seems to be be the 'eager' evaluation and discarding logic that I added in https://github.com/rust-lang/rust/pull/85868 (and that I've now removed from this PR). When there are a large number of projections with a large number of sub-obligations, this appears to blow up compilation time. Unfortunately, removing that logic appears to have a large *negative* impact on the `deeply-nested` benchmark.

**Comment by JohnCSimon** [comment]
Returning to author because this seems to require more conversation. This also has a merge conflict.

Adjust the label with `@rustbot ready` when its ready for review

**Comment by jackh726** [comment]
After #91065 lands (soon), I'd like to start moving on getting this (and other related PRs) landed.

I think first, we should try to land the deduplication done in #84944/#90423. The only *additional* thing that #84944 does that #90423 does is remove the deduplication in `select/mod.rs`, which seems to provide a moderate perf gain (though, the base for these changes are different, so it might make no different). I'm also looking at the crater run done in #77325 and wondering why we never made tests for the variants found there (or did we?)

After that change lands (which should be a big perf win), we should take another look at this PR (see how perf looks). But, I'm inclined to agree with https://github.com/rust-lang/rust/pull/91183#issuecomment-981127777 that we likely just need to accept the perf hit.

However, I do think it would be super helpful for a general writeup of how projection and projection caching works. I don't think it needs to block this PR, but it would be really nice. There are issues like #80807 and #87755 that still lurk that it would be really nice to have a good explanation of the current system to help get a better birds-eye view of the problem, in case that helps us see a solution.

**Comment by Aaron1011** [comment]
Note that https://github.com/rust-lang/rust/pull/91186 is another PR that does some form of deduplication

**Comment by Aaron1011** [comment]
cc @rust-lang/infra: I don't know why `T-rustdoc` got added

**Comment by wesleywiser** [comment]
I retested on an internal crate that was experiencing a huge regression in build times with the 1.56 release:

- On nightly: 898 seconds to build
- This patch: 34 seconds to build

Really great work! 🎉 🎉 

I'm planning to review again tomorrow sometime and r+ assuming everything looks good.

File: compiler/rustc_infer/src/traits/project.rs
Comment (jackh726) [inline compiler/rustc_infer/src/traits/project.rs:188]: So, the docs above say "During evaluation, we will skip evaluating the cached sub-obligations in `ty` if this field is set." So, setting the obligations here to an empty vec is a bit redundant?

Seemingly, at the very least, the docs should be changed to reflect that the obligations are removed altogether. But maybe we can improve the type here to be encode the either-or state or "obligations or completion"

File: compiler/rustc_trait_selection/src/lib.rs
Comment (jackh726) [reply compiler/rustc_trait_selection/src/lib.rs:15]: I think so too

File: compiler/rustc_trait_selection/src/traits/select/mod.rs
Comment (jackh726) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:597]: I feel like you should be able to get around this `unwrap` with an if-let and binding...

File: compiler/rustc_trait_selection/src/traits/fulfill.rs
Comment (jackh726) [inline compiler/rustc_trait_selection/src/traits/fulfill.rs:724]: Currently we can `complete` a projection in two places. I'm curious which of the two gives the perf benefits. It could be that one is essentially useless (in which case that might suggest that code path is not followed much - at least in terms where caching could be helpful).

Either way, I think I'd at least like some better comments about *why* we can cache at each place.

File: compiler/rustc_trait_selection/src/traits/project.rs
Comment (jackh726) [inline compiler/rustc_trait_selection/src/traits/project.rs:969]: I think I'd like to see this change pulled out to a separate PR - considering it has it's own bug fixes. Or at least have a separate PR filed to see the perf effects of removing this on its own.

**Review by jackh726 — CHANGES_REQUESTED** [review]
I have some things I'd like to see here for perf experimenting - but that doesn't necessarily need to happen.

There are some comments that I'd like changed though, after that r=me if @Aaron1011 doesn't want to do the perf experiments I commented on.

**Comment by Mark-Simulacrum** [comment]
Note that due to merging https://github.com/rust-lang/rustc-perf/pull/1123, these runs may look quite bad -- once we get another master commit landed + benchmarked, re-running perf should make any noise there go away. It shouldn't affect non-incremental results, though, of course.

(Leaving this comment on all pending perf run PRs).

File: compiler/rustc_infer/src/traits/project.rs
Comment (Aaron1011) [reply compiler/rustc_infer/src/traits/project.rs:188]: I think an empty `Vec` is fine, and makes the type a little easier to work with. I'll update the docs to explain what's going on.

**Comment by jackh726** [comment]
r=me with okay perf

**Comment by matze** [comment]
Any chance this can make it into 1.58? It fixes severe build times we have since 1.56 in a tonic/tower-based application of ours and essentially holding us back from upgrading to edition 2021 and a more recent rustls.

**Comment by jackh726** [comment]
I'll nominate this for stable and beta.

Personally, I think is *at least* as risky as #90423, and that got stable declined. But I'll beta backport might be okay here. I'll nominate both for discussion.

**Comment by pnkfelix** [comment]
Discussed in T-compiler meeting. Deferring decision on whether  to backport (any of beta/stable) until next week, at least.

**Comment by apiraino** [comment]
Discussed again today and the decision is to not baclport this on both beta or stable channels [Zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-12-30.20.2354818/near/266435406)

@rustbot label -beta-nominated -stable-nominated

**Pull request merged**
